### PR TITLE
LT-437/cache-expire-drain

### PR DESCRIPTION
**Pull request by danielcooper**
This Pr:

- Adds a ExpiringCache Drain. This is useful for clients that build their own view of the world and so do not need/want events proactively cached
- Adds ExpireCache middleware
- Adds ExpireCache middleware to the regular caching drain *before* the filtering. This ensure that we never store an old response in the following scenario:

1. event 1 arrives
2. dirty map flagged
3. cache invalidated
4. app notified
5. app fetches resource (v1 is cached)
6. event 2 arrives
7. dirty map already flagged
8. cache not invalidated
9. app fetches resource (v1 still cached)

### CODE REVIEW

File: lib/routemaster/drain/expiring_cache.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+module Routemaster
+  module Drain
+    # Rack application which authenticates, parses, filters duplicates in this request
+    # and invalidates the cache for all updated or new items.
+    #
+    # See the various corresponding middleware for details on operation:
+    # {Middleware::RootPostOnly}, {Middleware::Authenticate},
+    # {Middleware::Parse}, {Middleware::ExpireCache} and {Terminator}.
+    class ExpiringCache
```
Comment (mezis) [inline lib/routemaster/drain/expiring_cache.rb]: Naming: shouldn't this be `CacheExpiring`, as in "a middleware that expires the cache" (or even better `CacheBusting`)?

This reads as "a middleware that is or has an expiring cache" and seems a little confusing (to me!!)

File: lib/routemaster/drain/caching.rb [hunk 2]
```diff
@@ -28,6 +29,7 @@ def initialize(options = {})
           use Middleware::RootPostOnly
           use Middleware::Authenticate, options
           use Middleware::Parse
+          use Middleware::ExpireCache,  options
```
Comment (mezis) [inline lib/routemaster/drain/caching.rb:32]: Seeing as you're fixing this concurrency bug, you might want to pick up + close [LT-447](https://deliveroo.atlassian.net/browse/LT-447).

File: lib/routemaster/middleware/payload_filter.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module Routemaster
+  module Middleware
+    class PayloadFilter
+      # Filters duplicate events by url and type in a single payload.
+      def run(payload)
+        payload.group_by { |event| [event['url'], event['type']] }.map { |_, events| events.last }
```
Comment (mezis) [inline lib/routemaster/middleware/payload_filter.rb:6]: You're assuming order within a payload, which is not guaranteed by the bus.

File: spec/routemaster/drain/caching_spec.rb [hunk 4]
```diff
@@ -37,7 +38,7 @@
   it 'increments the event index' do
     ei_double = double(increment: 1)
     allow(Routemaster::EventIndex).to receive(:new).and_return(ei_double)
-    expect(ei_double).to receive(:increment).exactly(3).times
+    expect(ei_double).to receive(:increment).exactly(4).times
```
Comment (mezis) [inline spec/routemaster/drain/caching_spec.rb:41]: This looks like a magical value change. Why exactly 4 times?

File: lib/routemaster/middleware/payload_filter.rb (hunk 3 above)
Comment (danielcooper) [reply lib/routemaster/middleware/payload_filter.rb:6]: I don't think I am. `events.last` could just as easily be `events.first` and it wouldn't make a difference (apart from some very slight change in `t`)

File: lib/routemaster/drain/expiring_cache.rb (hunk 1 above)
Comment (danielcooper) [reply lib/routemaster/drain/expiring_cache.rb]: Sure, happy with those names

File: spec/routemaster/drain/caching_spec.rb (hunk 4 above)
Comment (danielcooper) [reply spec/routemaster/drain/caching_spec.rb:41]: Previously we didn't expire for every event (as we did it after the dirty map) - this removed one of the duplicate events from the spec. Now we expire for every event (but still initiate a de-duped cache step)

File: lib/routemaster/middleware/payload_filter.rb (hunk 3 above)
Comment (gregbeech) [reply lib/routemaster/middleware/payload_filter.rb:6]: Probably does make sense to return the latest `t` though, just to avoid any confusion.

**Review by gregbeech — APPROVED** [review]
One small comment, but generally LGTM.

**Pull request merged**